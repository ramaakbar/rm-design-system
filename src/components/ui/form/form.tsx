import { ComponentProps } from "react";
import {
  FieldValues,
  FormProvider,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { twMerge } from "tailwind-merge";

interface Props<T extends FieldValues>
  extends Omit<ComponentProps<"form">, "onSubmit"> {
  form: UseFormReturn<T>;
  onSubmit: SubmitHandler<T>;
}

const Form = <T extends FieldValues>({
  form,
  onSubmit,
  children,
  className,
  ...props
}: Props<T>) => (
  <FormProvider {...form}>
    <form onSubmit={form.handleSubmit(onSubmit)} {...props}>
      <fieldset
        disabled={form.formState.isSubmitting}
        className={twMerge(className)}
      >
        {children}
      </fieldset>
    </form>
  </FormProvider>
);

export default Form;
