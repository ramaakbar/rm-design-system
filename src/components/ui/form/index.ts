export * from "./checkbox";
export * from "./nativeSelect";
export { PasswordInput } from "./passwordInput";
export { TextArea } from "./textArea";
export { TextInput } from "./textInput";
export * from "./radio";
export * from "./form";
export * from "./inputWrapper";
