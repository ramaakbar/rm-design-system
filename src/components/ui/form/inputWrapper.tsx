import { twMerge } from "tailwind-merge";

type InputWrapperProps = {
  name: string;
  label?: string;
  required?: boolean;
  description?: string;
  error?: string;
} & React.PropsWithChildren;

export function InputWrapper({
  name,
  label,
  required,
  description,
  error,
  children,
}: InputWrapperProps) {
  return (
    <div className="flex flex-col">
      <label htmlFor={name} className="text-sm font-medium text-foreground">
        {label && label}
        {label && required && <span className="text-destructive"> *</span>}
        {description && <p className="text-muted-foreground">{description}</p>}
      </label>
      <div
        className={twMerge(
          "relative",
          (label || description) && "mt-1.5",
          error && "mb-1.5"
        )}
      >
        {children}
      </div>
      {error && <span className="text-sm text-destructive">{error}</span>}
    </div>
  );
}
