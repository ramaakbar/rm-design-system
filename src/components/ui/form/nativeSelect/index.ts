export * from "./nativeSelect";
export * from "./nativeSelect.stories";
