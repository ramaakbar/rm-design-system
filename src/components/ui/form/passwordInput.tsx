import { forwardRef, useState } from "react";
import { Eye, EyeOff } from "lucide-react";
import { useFormContext } from "react-hook-form";
import { twMerge } from "tailwind-merge";

import { TextInputProps } from "./textInput";

const PasswordInput = forwardRef<HTMLInputElement, TextInputProps>(
  (
    {
      label,
      placeholder,
      description,
      required,
      disabled = false,
      startNode: StartNode,
      endNode: EndNode,
      ...props
    },
    ref
  ) => {
    const [showPassword, setShowPassword] = useState(false);
    const form = useFormContext();
    const state = form.getFieldState(props.name, form.formState);

    const handleTogglePassword = () => setShowPassword((prev) => !prev);

    return (
      <div className="flex flex-col">
        <label htmlFor={props.name} className="text-sm font-medium">
          {label && label}
          {label && required && <span className="text-destructive"> *</span>}
          {description && (
            <p className="text-muted-foreground">{description}</p>
          )}
        </label>
        <div
          className={twMerge(
            "relative",
            (label || description) && "mt-1.5",
            state.error && "mb-1.5"
          )}
        >
          {StartNode && (
            <div className="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3">
              {typeof StartNode === "string" ? (
                <span className="text-muted-foreground">{StartNode}</span>
              ) : (
                <StartNode className="h-4 w-4 text-muted-foreground" />
              )}
            </div>
          )}
          <input
            type={showPassword ? "text" : "password"}
            id={props.name}
            ref={ref}
            disabled={disabled}
            readOnly={disabled}
            placeholder={placeholder}
            required={required}
            className={twMerge(
              "block w-full rounded-md border-0 bg-transparent px-3 py-2 ring-1 ring-inset ring-input placeholder:text-muted-foreground focus:ring-2 focus:ring-inset focus:ring-indigo-600 disabled:cursor-not-allowed disabled:bg-muted disabled:opacity-50",
              state.error &&
                "text-destructive ring-destructive focus:ring-destructive/80",
              StartNode && "pl-10",
              EndNode && "pr-10"
            )}
            {...props}
          />

          <div
            className="absolute inset-y-0 right-0 flex cursor-pointer select-none items-center pr-3"
            onClick={handleTogglePassword}
          >
            {!showPassword ? (
              <Eye className="h-4 w-4 text-muted-foreground" />
            ) : (
              <EyeOff className="h-4 w-4 text-muted-foreground" />
            )}
          </div>
        </div>
        {state.error && (
          <span className="text-sm text-destructive">
            {state.error.message}
          </span>
        )}
      </div>
    );
  }
);

PasswordInput.displayName = "PasswordInput";

export default PasswordInput;
