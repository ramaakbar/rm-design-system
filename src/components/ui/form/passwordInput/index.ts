export * from "./passwordInput";
export * from "./passwordInput.stories";
