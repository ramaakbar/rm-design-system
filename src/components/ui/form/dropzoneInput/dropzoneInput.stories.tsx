import type { Meta, StoryObj } from "@storybook/react";
import { z } from "zod";

import { useZodForm } from "../../../../hooks/useZodForm";
import { Form } from "../form";
import { DropzoneInput } from "./dropzoneInput";

const meta = {
  title: "UI/DropzoneInput",
  component: DropzoneInput,
  tags: ["autodocs"],
  args: {
    name: "name",
    label: "Label",
    disabled: false,
  },
} satisfies Meta<typeof DropzoneInput>;

export default meta;
type Story = StoryObj<typeof meta>;

const schema = z.object({
  picture: z.array(z.custom<File>((v) => v instanceof File)).optional(),
});

const DropzoneExample = () => {
  const form = useZodForm({
    schema,
  });

  return (
    <Form form={form} onSubmit={() => {}}>
      <DropzoneInput name="asdasd" />
    </Form>
  );
};

export const Input: Story = {
  render: () => <DropzoneExample />,
};
