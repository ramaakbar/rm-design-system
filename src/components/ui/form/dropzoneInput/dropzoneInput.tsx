import {
  ComponentProps,
  forwardRef,
  useCallback,
  useEffect,
  useState,
} from "react";
import { Image as ImageIcon, X } from "lucide-react";
import {
  Accept,
  FileRejection,
  FileWithPath,
  useDropzone,
} from "react-dropzone";
import { Controller, useFormContext } from "react-hook-form";
import { twMerge } from "tailwind-merge";

import { Button } from "../../button";
import { InputWrapper } from "../inputWrapper";

type DropzoneInputProps = Omit<ComponentProps<"input">, "accept"> & {
  label?: string;
  name: string;
  description?: string;
  error?: string;
  maxFiles?: number;
  maxSize?: number;
  accept?: Accept;
};

export const DropzoneInput = forwardRef<HTMLInputElement, DropzoneInputProps>(
  (
    {
      label,
      description,
      required,
      disabled = false,
      error,
      maxFiles = 1,
      maxSize = 2 * 1000 * 1000,
      accept,
      className,
      ...props
    },
    ref
  ) => {
    const { control, getValues, setValue, watch, setError } = useFormContext();
    const [files, setFiles] = useState<FileWithPath[]>(watch(props.name));

    const onDrop = useCallback(
      <T extends File>(acceptedFiles: T[], rejectedFiles: FileRejection[]) => {
        if (rejectedFiles && rejectedFiles.length > 0) {
          setValue(props.name, []);
          setFiles([]);
          setError(props.name, {
            type: "manual",
            message: rejectedFiles && rejectedFiles[0].errors[0].message,
          });
        } else {
          setFiles(acceptedFiles);
          if (maxFiles > 1) {
            setValue(props.name, acceptedFiles, { shouldValidate: true });
          } else {
            setValue(props.name, acceptedFiles[0], { shouldValidate: true });
          }
        }
      },
      [props.name, setValue, setError, maxFiles]
    );

    useEffect(() => {
      if (getValues(props.name) === undefined) setFiles([]);

      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [getValues(props.name)]);

    const removeFile = (index: FileWithPath) => {
      const fileCopy = [...files];
      const removedFile = fileCopy.indexOf(index);
      if (removedFile > -1) {
        fileCopy.splice(removedFile, 1);
      }
      setValue(props.name, fileCopy);
      setFiles(fileCopy);
    };

    const { getRootProps, getInputProps, open } = useDropzone({
      onDrop,
      noClick: true,
      noKeyboard: true,
      maxFiles: maxFiles,
      maxSize: maxSize,
      accept: accept,
    });

    return (
      <InputWrapper
        label={label}
        description={description}
        required={required}
        name={props.name}
        error={error}
      >
        {files?.length > 0 ? (
          <div
            className={twMerge(
              "block w-full rounded-md border-0 bg-transparent px-3 py-2 ring-1 ring-inset ring-input placeholder:text-muted-foreground focus:ring-2 focus:ring-inset focus:ring-indigo-600 disabled:cursor-not-allowed disabled:bg-muted disabled:opacity-50"
            )}
          >
            {files?.map((file, i) => (
              <div key={i} className="flex w-full items-center justify-between">
                <p>{file.path}</p>
                <Button
                  variant="ghost"
                  className="text-muted-foreground"
                  size="sm"
                  onClick={() => removeFile(file)}
                >
                  <X size={20} />
                </Button>
              </div>
            ))}
          </div>
        ) : (
          <Controller
            name={props.name}
            control={control}
            defaultValue=""
            render={({ field }) => (
              <div
                {...getRootProps()}
                className={twMerge(
                  "block w-full cursor-pointer rounded-md border border-dashed border-border bg-transparent px-6 py-10 ring-0 ",
                  error && "border-destructive",
                  disabled && "cursor-not-allowed bg-muted opacity-50",
                  className
                )}
                {...field}
                onClick={open}
              >
                <div className="flex flex-col items-center justify-center">
                  <ImageIcon
                    className="mb-5 opacity-80"
                    size={50}
                    strokeWidth={1}
                  />
                  <div>
                    <button type="button" className="text-indigo-600">
                      Click here to upload a file
                    </button>{" "}
                    or drag and drop
                  </div>
                  <input
                    id={props.name}
                    disabled={disabled}
                    readOnly={disabled}
                    required={required}
                    ref={ref}
                    {...getInputProps()}
                  />
                </div>
              </div>
            )}
          />
        )}
      </InputWrapper>
    );
  }
);

DropzoneInput.displayName = "DropzoneInput";
