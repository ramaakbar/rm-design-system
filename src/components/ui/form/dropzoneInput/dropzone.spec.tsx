import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useForm } from "react-hook-form";

import { Form } from "../form";
import { DropzoneInput } from "./dropzoneInput";

describe("DropzoneInput", () => {
  const defaultProps = {
    name: "fileInput",
    label: "Upload File",
    accept: {
      "image/*": [],
    },
  };

  const FormTest = () => {
    const form = useForm();
    return (
      <Form form={form} onSubmit={() => {}}>
        <DropzoneInput {...defaultProps} />
      </Form>
    );
  };

  it("should render properly", () => {
    const { container } = render(<FormTest />);
    expect(container).toMatchSnapshot();
  });

  it("should display the label correctly", () => {
    render(<FormTest />);
    const label = screen.getByText("Upload File");
    expect(label).toBeInTheDocument();
  });

  it("should display the file path after selecting a file", async () => {
    render(<FormTest />);

    const input = screen.getByLabelText("Upload File");
    const file = new File(["test file content"], "test.png", {
      type: "image/png",
    });
    userEvent.upload(input, file);

    const filePath = await screen.findByText("test.png");
    expect(filePath).toBeInTheDocument();
  });

  it("should allow user to remove a selected file", async () => {
    render(<FormTest />);

    const input = screen.getByLabelText("Upload File");
    const file = new File(["test file content"], "test.png", {
      type: "image/png",
    });
    await userEvent.upload(input, file);

    const removeButton = screen.getByRole("button");
    await userEvent.click(removeButton);

    const dropzoneText = await screen.findByText("Click here to upload a file");

    expect(dropzoneText).toBeInTheDocument();
  });
});
