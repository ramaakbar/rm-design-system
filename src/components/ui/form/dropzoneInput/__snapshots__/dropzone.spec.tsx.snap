// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropzoneInput should render properly 1`] = `
<div>
  <form>
    <fieldset
      class=""
    >
      <div
        class="flex flex-col"
      >
        <label
          class="text-sm font-medium text-foreground"
          for="fileInput"
        >
          Upload File
        </label>
        <div
          class="relative mt-1.5"
        >
          <div
            class="block w-full cursor-pointer rounded-md border border-dashed border-border bg-transparent px-6 py-10 ring-0"
            name="fileInput"
            role="presentation"
            value=""
          >
            <div
              class="flex flex-col items-center justify-center"
            >
              <svg
                class="mb-5 opacity-80"
                fill="none"
                height="50"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1"
                viewBox="0 0 24 24"
                width="50"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <circle
                  cx="9"
                  cy="9"
                  r="2"
                />
                <path
                  d="m21 15-3.086-3.086a2 2 0 0 0-2.828 0L6 21"
                />
              </svg>
              <div>
                <button
                  class="text-indigo-600"
                  type="button"
                >
                  Click here to upload a file
                </button>
                 
                or drag and drop
              </div>
              <input
                accept="image/*"
                id="fileInput"
                multiple=""
                style="display: none;"
                tabindex="-1"
                type="file"
              />
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </form>
</div>
`;
